integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/JtaConnection.java

```diff
@@ -1124,27 +1124,18 @@ private Connection connectionFunction(Xid xid) {
     private void transactionCompleted(int commitedOrRolledBack) {
         this.enlistment = null; // volatile write
         try {
-            if (!super.isClosed()) {
-                // Did someone call close() while we were enlisted?  That's permitted by section 4.2 of the JTA
-                // specification, but it shouldn't actually close the connection because the prepare/commit cycle
-                // wouldn't have started.
-                boolean closeWasPending = this.isClosePending(); // volatile read
-
-                // If they did, then we must be non-closeable. If they didn't, we could be either closeable or not
-                // closeable.
-                assert !closeWasPending || !this.isCloseable();
-
-                // Now the global transaction is over, so blindly set our closeable status to true. (It may already be
-                // true.)  This resets the closePending status, per spec.
-
-                this.setCloseable(true);
-                assert this.isCloseable();
+            boolean closeWasPending = this.isClosePending();
+            this.setCloseable(true);
+            assert this.isCloseable();
+            assert !this.isClosePending();
+            if (closeWasPending) {
+                assert !this.isClosed();
+                assert !this.delegate().isClosed();
+                this.close();
```

Comment:
Probably inconsequential, but the `close` method in this class (I did not look at `super.close()` does not check `isCloseable()`. In fact, in this class only the `assert`s here check it. Is invoking `setCloseable(true)` before invoking `close()` necessary? Redundant? Immaterial? (I guess I'm mostly just grasping for the underlying state model represented by the various flags.)

Reply:
Fair questions all.
1. `super.close()` (`ConditionallyCloseableConnection#close()`) does in fact check `isCloseable()`, and takes different actions depending on the outcome. If the connection is closeable, then the delegate it wraps gets its `close()` method called. If the connection is not closeable (because it is enrolled in a transaction, for example), then the intent to close it is recorded and action is taken at transaction completion time to actualize the request.
2. Calling `setCloseable(true)` here is critical. Basically, someone called `close()` on the `JtaConnection`, and, now that the transaction is definitely completed, that operation that the caller _thought_ was immediate but was not should _now_ take effect. It can only take effect if the `JtaConnection` is, in fact, closeable.